Add validation checks for GS attributes after parsing

antiagainst opened the pull request:
This commit add checking for two GS rules:

* GS entry point should have the maxvertexcount attribute
* Stream-output objects should be inout parameters

antiagainst commented:
Right now we have assertion failures if the `maxvertexcount` on entry point or `inout` on stream-output parameters in geometry shader is missing. This commit fixes these two issues.

To properly check and error out for the above two cases, we need the shader model and entry point. I was thinking to put the check into `hlsl::DiagnoseTranslationUnit()`, which pinpoints the entry point, but it seems the current DXIL codegen depends on `HandleTopLevelDecl()`. So that won't fix the assertion failures.

Thus, separated check after each top-level decl. This causes a new `RecursiveASTVisitor` to be created for each top-level decl. Slightly performance penalty, and different behavior compared to `fxc`. When the entry function is redefined, we will output the redefinition error together with all attribute errors on all duplicated entry points, while `fxc` only outputs redefinition error. `fxc` only checks `inout` on functions eventually got called by the entry point, but we check all of them. However, IMHO, these two mismatches are not big problems.

I saw there are TODOs in `ConstructFieldAttributedAnnotation()` saying `validation this at ASTContext::getFunctionType in AST/ASTContext.cpp`. But I don't know how that can be achieved.

Let me know if this is not the preferred way of validating these issues. Happy to adjust. :)g

tex3d commented on tools/clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -8882,6 +8883,84 @@ static NameLookup GetSingleFunctionDeclByName(clang::Sema *self, StringRef Name,
+  // Validate that stream-ouput objects are marked as inout
+  if (shaderModel->IsGS()) {
+    if (hlsl::IsHLSLStreamOutputType(parmVar->getType()) &&
+        !parmVar->hasAttr<HLSLInOutAttr>()) {
```
Perhaps this validation could be applied without regard to the shader model.  Otherwise, it's being applied to all functions, but only if the current entry is a GS.

tex3d replied:
Checking additional shader model specific requirements might be more difficult, since we will need to know whether this function is considered an entry, and which kind (esp. for lib target).

tex3d replied:
I think this particular check could be globally applied in Sema::DiagnoseHLSLDecl.  Just need to move some HLSL type retrieval code out of if (hasSignSpec), then just before the end of the method, verify.
Here's a start:
```
diff --git a/tools/clang/lib/Sema/SemaHLSL.cpp b/tools/clang/lib/Sema/SemaHLSL.cpp
index 53d1f23d..9801de8e 100644
--- a/tools/clang/lib/Sema/SemaHLSL.cpp
+++ b/tools/clang/lib/Sema/SemaHLSL.cpp
@@ -10551,6 +10551,10 @@ bool Sema::DiagnoseHLSLDecl(Declarator &D, DeclContext *DC,
     }
   }

+  HLSLExternalSource *hlslSource = HLSLExternalSource::FromSema(this);
+  ArTypeObjectKind objKind = hlslSource->GetTypeObjectKind(qt);
+  ArBasicKind basicKind = hlslSource->GetTypeElementKind(qt);
+
   // Check for deprecated effect object type here, warn, and invalidate decl
   bool bDeprecatedEffectObject = false;
   bool bIsObject = false;
@@ -10609,9 +10613,6 @@ bool Sema::DiagnoseHLSLDecl(Declarator &D, DeclContext *DC,
   }

   if (hasSignSpec) {
-     HLSLExternalSource *hlslSource = HLSLExternalSource::FromSema(this);
-     ArTypeObjectKind objKind = hlslSource->GetTypeObjectKind(qt);
-     ArBasicKind basicKind = hlslSource->GetTypeElementKind(qt);
      // vectors or matrices can only have unsigned integer types.
      if (objKind == AR_TOBJ_MATRIX || objKind == AR_TOBJ_VECTOR || objKind == AR_TOBJ_BASIC || objKind == AR_TOBJ_ARRAY) {
          if (!IS_BASIC_UNSIGNABLE(basicKind)) {
@@ -10889,6 +10890,15 @@ bool Sema::DiagnoseHLSLDecl(Declarator &D, DeclContext *DC,
     }
   }

+  if (isParameter
+      && ( basicKind == ArBasicKind::AR_OBJECT_LINESTREAM ||
+           basicKind == ArBasicKind::AR_OBJECT_POINTSTREAM ||
+           basicKind == ArBasicKind::AR_OBJECT_TRIANGLESTREAM )
+      && !(usageIn && usageOut)) {
+    // TODO: diagnose
+    result = false;
+  }
+
   if (!result) {
     D.setInvalidType();
   }
```

antiagainst replied:
Good point! Thanks for pointing me to `DiagnoseHLSLDecl()`. Didn't notice it previously. Definitely much suitable to check there.

tex3d commented on tools/clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -8882,6 +8883,84 @@ static NameLookup GetSingleFunctionDeclByName(clang::Sema *self, StringRef Name,
+bool HLSLAttributeChecker::VisitFunctionDecl(FunctionDecl *func) {
+  isEntryFunction = theSema.getLangOpts().HLSLEntryFunction == func->getName();
+
+  if (isEntryFunction && shaderModel->IsGS()) {
```
This is definitely complicated by the library case, and for patch constant function case if we add validation for that.

tex3d replied:
Seems the maxvertexcount check can easily be made in DiagnoseTranslationUnit (given the HLSLProfile addition).

antiagainst replied:
Yes. Moved the check here.

antiagainst commented:
Hey @tex3d, I've addressed your comments. PTAL.

Also, I added the third check for missing HS patch constant function. It also causes assertion failures if `patchconstantfunc` is missing. However, I have to do the check twice since DXIL does `HandleTopLevelDecl()`, which is earlier than `DiagnoseTranslationUnit()`. SPIR-V codegen relies on `DiagnoseTranslationUnit()`.
